tgui/packages/tgui/interfaces/Pandemic.tsx

```diff
@@ -0,0 +1,484 @@
+  viruses?: Virus[];
+};
+
+type Beaker = {
+  volume: number;
+  capacity: number;
+};
+
+type Blood = {
+  dna: string;
+  type: string;
+};
+
+type Resistance = {
+  id: string;
+  name: string;
+};
+
+type Virus = {
+  name: string;
+  can_rename: number;
+  is_adv: number;
+  symptoms: Symptom[];
+  resistance: number;
+  stealth: number;
+  stage_speed: number;
+  transmission: number;
+  index: number;
+  agent: string;
+  description: string;
+  spread: string;
+  cure: string;
+};
+
+type VirusInfoProps = {
+  virus: Virus;
+};
+
+type Symptom = {
+  name: string;
+  desc: string;
+  stealth: number;
+  resistance: number;
+  stage_speed: number;
+  transmission: number;
+  level: number;
+  neutered: number;
+  threshold_desc: Threshold[];
+};
+
+type SymptomDisplayProps = {
+  symptoms: Symptom[];
+};
+
+type Threshold = {
+  label: string;
+  descr: string;
+};
+
+type ThresholdDisplayProps = {
+  thresholds: Threshold[];
+};
+
+export const Pandemic = (_, context) => {
+  const { data } = useBackend<PandemicContext>(context);
+  const { has_beaker, has_blood } = data;
+
+  return (
+    <Window width={650} height={500}>
+      <Window.Content>
+        <Stack fill vertical>
+          <Stack.Item>
+            <BeakerDisplay />
+          </Stack.Item>
+          {!!has_beaker && !!has_blood && (
+            <Stack.Item grow>
+              <SpecimenDisplay />
+            </Stack.Item>
+          )}
+        </Stack>
+      </Window.Content>
+    </Window>
+  );
+};
+
+/** Displays loaded container info, if it exists */
+const BeakerDisplay = (_, context) => {
+  const { act, data } = useBackend<PandemicContext>(context);
+  const { has_beaker, beaker, has_blood } = data;
+  const cant_empty = !has_beaker || !beaker?.volume;
+
+  return (
+    <Section
+      title="Beaker"
+      buttons={
+        <>
+          <Button
+            icon="times"
+            content="Empty and Eject"
+            color="bad"
+            disabled={cant_empty}
+            onClick={() => act('empty_eject_beaker')}
+          />
+          <Button
+            icon="trash"
+            content="Empty"
+            disabled={cant_empty}
+            onClick={() => act('empty_beaker')}
+          />
+          <Button
+            icon="eject"
+            content="Eject"
+            disabled={!has_beaker}
+            onClick={() => act('eject_beaker')}
+          />
+        </>
+      }>
+      {has_beaker ? (
+        beaker?.volume ? (
+          has_blood ? (
+            <Stack vertical>
+              <Stack.Item>
+                <BeakerInfoDisplay />
+              </Stack.Item>
+              <Stack.Item>
+                <AntibodyInfoDisplay />
+              </Stack.Item>
+            </Stack>
+          ) : (
+            <NoticeBox>No blood detected</NoticeBox>
+          )
+        ) : (
+          <NoticeBox>Beaker is empty</NoticeBox>
+        )
+      ) : (
+        <NoticeBox>No beaker loaded</NoticeBox>
+      )}
```

Comment:
Can this be re-written without ternaries?
Something like
```
(!has_beaker && (<NoticeBox>...</NoticeBox>)) ||
(!beaker.volume && (<NoticeBox>...</NoticeBox>)) ||
(!has_blood && (<NoticeBox>...</NoticeBox>)) ||
(<Stack>...</Stack>)
```
This is the equivalent of `if(!has_beaker) {...} else if (!beaker.volume) {...} else if (!has blood) {...} else {...}`

React has short-circuiting evaluations, so if `!has_beaker` evaluates to truthy, then `(!has_beaker && (<NoticeBox>...</NoticeBox>))` will evaluate to simply `(<NoticeBox>...</NoticeBox>)` (which is also truthy) and short-circuit, not performing any of the other logic because `true || ...` will always evaluate to true. The same holds true for the next if there is a beaker but it has nothing in it, etc.

For more info google `react short circuit evaluation`.

Reply:
for readability purposes I can make this into a simpler ```content``` variable. will do this soon, let me know if you prefer it
i think the nested ternary was just the previous logic and I didn't alter it

Reply:
this is done, if you're not happy with it I can rewrite it in short circuit also. i feel people get confused by larger short circuit expressions so it's condensed to a simple if/else variable